store/v2/multi/snapshot.go

```diff
@@ -0,0 +1,232 @@
+package root
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	prefixdb "github.com/cosmos/cosmos-sdk/db/prefix"
+	"github.com/cosmos/cosmos-sdk/snapshots"
+	snapshottypes "github.com/cosmos/cosmos-sdk/snapshots/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	types "github.com/cosmos/cosmos-sdk/store/v2"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Snapshot implements snapshottypes.Snapshotter.
+func (rs *Store) Snapshot(height uint64, format uint32) (<-chan io.ReadCloser, error) {
+	if format != snapshottypes.CurrentFormat {
+		return nil, sdkerrors.Wrapf(snapshottypes.ErrUnknownFormat, "format %v", format)
+	}
+
+	if height == 0 {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "cannot snapshot height 0")
+	}
+	if height > uint64(rs.LastCommitID().Version) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrLogic, "cannot snapshot future height %v", height)
+	}
+	versions, err := rs.stateDB.Versions()
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "error while getting the snapshot versions at height %v", height)
+	}
+	if !versions.Exists(height) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "cannot find snapshot at height %v", height)
+	}
+
+	// get the saved snapshot at height
+	vs, err := rs.getView(int64(height))
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, fmt.Sprintf("error while get the version at height %d", height))
+	}
+
+	// Spawn goroutine to generate snapshot chunks and pass their io.ReadClosers through a channel
+	ch := make(chan io.ReadCloser)
+	go func() {
+		// setup snapshot export stream
+		chunkWriter, bufWriter, zWriter, protoWriter, err := snapshots.SetupExportStreamPipeline(ch)
+		if err != nil {
+			return
+		}
+		defer chunkWriter.Close()
+		defer func() {
+			if err := bufWriter.Flush(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		defer func() {
+			if err := zWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		defer func() {
+			if err := protoWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
```

Comment:
this is ugly, here we have other implementation: https://github.com/cosmos/cosmos-sdk/pull/10961